test-common/src/main/scala/org/scalajs/testcommon/Serializer.scala

```diff
@@ -0,0 +1,290 @@
+        out.put(fp.isModule)
+        out.put(fp.superclassName)
+        out.put(fp.requireNoArgConstructor)
+      case _ =>
+        throw new IllegalArgumentException(
+            s"Unknown Fingerprint type: ${fp.getClass}")
+    }
+
+    def deserialize(in: DeserializeState): Fingerprint = in.s.readByte() match {
+      case ANNOTATED =>
+        new AnnotatedFingerprint {
+          val isModule: Boolean = in.get[Boolean]()
+          val annotationName: String = in.get[String]()
+        }
+
+      case SUBCLASS =>
+        new SubclassFingerprint {
+          val isModule: Boolean = in.get[Boolean]()
+          val superclassName: String = in.get[String]()
+          val requireNoArgConstructor: Boolean = in.get[Boolean]()
+        }
+
+      case t =>
+        throw new IOException(s"Unknown Fingerprint type: $t")
+    }
+  }
+
+  implicit object SelectorSerializer extends Serializer[Selector] {
+    // Type tags.
+    private val SUITE: Byte = 1
+    private val TEST: Byte = 2
+    private val NESTED_SUITE: Byte = 3
+    private val NESTED_TEST: Byte = 4
+    private val TEST_WILDCARD: Byte = 5
+
+    def serialize(sel: Selector, out: SerializeState): Unit = sel match {
+      case sel: SuiteSelector => out.s.writeByte(SUITE)
+
+      case sel: TestSelector =>
+        out.s.writeByte(TEST)
+        out.put(sel.testName)
+
+      case sel: NestedSuiteSelector =>
+        out.s.writeByte(NESTED_SUITE)
+        out.put(sel.suiteId)
+
+      case sel: NestedTestSelector =>
+        out.s.writeByte(NESTED_TEST)
+        out.put(sel.suiteId)
+        out.put(sel.testName)
+
+      case sel: TestWildcardSelector =>
+        out.s.writeByte(TEST_WILDCARD)
+        out.put(sel.testWildcard)
+
+      case _ =>
+        throw new IllegalArgumentException(
+            s"Unknown Selector type: ${sel.getClass}")
+    }
+
+    def deserialize(in: DeserializeState): Selector = in.s.readByte() match {
+      case SUITE         => new SuiteSelector()
+      case TEST          => new TestSelector(in.get[String]())
+      case NESTED_SUITE  => new NestedSuiteSelector(in.get[String]())
+      case NESTED_TEST   => new NestedTestSelector(in.get[String](), in.get[String]())
+      case TEST_WILDCARD => new TestWildcardSelector(in.get[String]())
+      case t             => throw new IOException(s"Unknown Selector type: $t")
+    }
+  }
+
+  implicit object TaskDefSerializer extends Serializer[TaskDef] {
+    def serialize(x: TaskDef, out: SerializeState): Unit = {
+      out.put(x.fullyQualifiedName)
+      out.put(x.fingerprint)
+      out.put(x.explicitlySpecified)
+      out.put(x.selectors.toList)
+    }
+
+    def deserialize(in: DeserializeState): TaskDef = {
+      new TaskDef(in.get[String](), in.get[Fingerprint](), in.get[Boolean](),
+          in.get[List[Selector]]().toArray)
+    }
+  }
+
+  implicit object StatusSerializer extends Serializer[Status] {
+    def serialize(x: Status, out: SerializeState): Unit =
+      out.s.writeInt(x.ordinal)
```

Comment:
I made a distinction between "data" primitive values and "control" primitive values. Maybe not a good idea?